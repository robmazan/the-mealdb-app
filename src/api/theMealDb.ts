const API_KEY = process.env.REACT_APP_MEALDB_API_KEY || 1;

export interface Meal {
  idMeal: number;
  strMeal: string;
  strDrinkAlternate: string;
  strCategory: string;
  strArea: string;
  strInstructions: string;
  strMealThumb: string;
  strTags: string;
  strYoutube: string;
  strIngredient1: string;
  strIngredient2: string;
  strIngredient3: string;
  strIngredient4: string;
  strIngredient5: string;
  strIngredient6: string;
  strIngredient7: string;
  strIngredient8: string;
  strIngredient9: string;
  strIngredient10: string;
  strIngredient11: string;
  strIngredient12: string;
  strIngredient13: string;
  strIngredient14: string;
  strIngredient15: string;
  strIngredient16: string;
  strIngredient17: string;
  strIngredient18: string;
  strIngredient19: string;
  strIngredient20: string;
  strMeasure1: string;
  strMeasure2: string;
  strMeasure3: string;
  strMeasure4: string;
  strMeasure5: string;
  strMeasure6: string;
  strMeasure7: string;
  strMeasure8: string;
  strMeasure9: string;
  strMeasure10: string;
  strMeasure11: string;
  strMeasure12: string;
  strMeasure13: string;
  strMeasure14: string;
  strMeasure15: string;
  strMeasure16: string;
  strMeasure17: string;
  strMeasure18: string;
  strMeasure19: string;
  strMeasure20: string;
  strSource: string;
  dateModified: string;
}

export interface Category {
  idCategory: number;
  strCategory: string;
  strCategoryThumb: string;
  strCategoryDescription: string;
}

export const searchByName = (name: string): Promise<Meal[]> =>
  fetch(
    `https://www.themealdb.com/api/json/v1/${API_KEY}/search.php?s=${encodeURIComponent(
      name
    )}`
  )
    .then(res => res.json())
    .then(res => res.meals);

export const searchByFirstLetter = (firstLetter: string): Promise<Meal[]> =>
  fetch(
    `https://www.themealdb.com/api/json/v1/${API_KEY}/search.php?f=${encodeURIComponent(
      firstLetter
    )}`
  )
    .then(res => res.json())
    .then(res => res.meals);

export const findById = (id: number): Promise<Meal> =>
  fetch(
    `https://www.themealdb.com/api/json/v1/${API_KEY}/lookup.php?i=${encodeURIComponent(
      id
    )}`
  )
    .then(res => res.json())
    .then(res => res.meals[0]);

export const fetchRandom = (): Promise<Meal> =>
  fetch(`https://www.themealdb.com/api/json/v1/${API_KEY}/random.php`)
    .then(res => res.json())
    .then(res => res.meals[0]);

export const fetchCategories = (): Promise<Category[]> =>
  fetch(`https://www.themealdb.com/api/json/v1/${API_KEY}/categories.php`)
    .then(res => res.json())
    .then(res => res.categories);
