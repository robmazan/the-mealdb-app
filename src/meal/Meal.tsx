import React, { useEffect, useState } from "react";
import { useParams } from "react-router";
import { findById } from "../api/theMealDb";
import { NormalizedMeal, normalizeMeal } from "../shared/tools";
import styles from "./Meal.module.scss";
import { Link } from "react-router-dom";

const Meal: React.FC = () => {
  const { mealId } = useParams();
  const [meal, setMeal] = useState<NormalizedMeal>();

  useEffect(() => {
    const fetchMeal = async () => {
      if (mealId) {
        const res = await findById(Number.parseInt(mealId));
        setMeal(normalizeMeal(res));
      }
    };
    fetchMeal();
  }, [mealId]);

  if (meal) {
    return (
      <>
        <Link className={styles.backLink} to={`/category/${meal.strCategory}`}>
          Back to category
        </Link>
        <h2 className={styles.pageTitle}>{meal.strMeal}</h2>
        <div className={styles.card}>
          <img
            className={styles.photo}
            src={meal.strMealThumb}
            alt={meal.strMeal}
          />
        </div>
        <h3>Ingredients</h3>
        <ul className={styles.ingredientList}>
          {meal.ingredients.map((ingredient, index) => (
            <li className={styles.listItem} key={`ing_${meal.idMeal}_${index}`}>
              {ingredient.measure}&nbsp;{ingredient.name}
            </li>
          ))}
        </ul>
        <ul className={styles.ingredientImages}>
          {meal.ingredients.map((ingredient, index) => (
            <li
              className={styles.listItem}
              key={`iim_${meal.idMeal}_${index}`}
              title={ingredient.name}
            >
              <img
                alt={ingredient.name}
                src={`https://www.themealdb.com/images/ingredients/${ingredient.name}-Small.png`}
              />
            </li>
          ))}
        </ul>
        <h3>Instructions</h3>
        {meal.instructions.length === 1 ? (
          <p>{meal.instructions[0]}</p>
        ) : (
          <ol>
            {meal.instructions.map((paragraph, index) => (
              <li
                className={styles.listItem}
                key={`ins_${meal.idMeal}_${index}`}
              >
                {paragraph}
              </li>
            ))}
          </ol>
        )}
      </>
    );
  } else {
    return <div>Loading...</div>;
  }
};

export default Meal;
