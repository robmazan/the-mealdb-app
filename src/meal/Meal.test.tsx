import { GlobalWithFetchMock } from "jest-fetch-mock";
import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import * as api from "../api/theMealDb";
import Meal from "./Meal";

const customGlobal: GlobalWithFetchMock = global as GlobalWithFetchMock;
customGlobal.fetch = require("jest-fetch-mock");
const fetchMock = customGlobal.fetch;

let mockMealId: string | undefined;
jest.mock("react-router", () => ({
  useParams() {
    return { mealId: mockMealId };
  }
}));

jest.mock("react-router-dom", () => ({
  Link: jest
    .fn()
    .mockImplementation(props => <code>{JSON.stringify(props)}</code>)
}));

let container: HTMLElement | null = null;
beforeEach(() => {
  fetchMock.resetMocks();
  mockMealId = "42";
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  if (container) {
    unmountComponentAtNode(container);
  }
  container!.remove();
  container = null;
});

it("renders the meal details properly", async () => {
  const fakeResponse: api.Meal = require("./__mocks__/mealDetails.json");
  fetchMock.mockResponseOnce(JSON.stringify(fakeResponse));

  await act(async () => {
    render(<Meal />, container);
  });

  expect(container!.innerHTML).toMatchInlineSnapshot(
    `"<code>{\\"className\\":\\"backLink\\",\\"to\\":\\"/category/Chicken\\",\\"children\\":\\"Back to category\\"}</code><h2 class=\\"pageTitle\\">Teriyaki Chicken Casserole</h2><div class=\\"card\\"><img class=\\"photo\\" src=\\"https://www.themealdb.com/images/media/meals/wvpsxx1468256321.jpg\\" alt=\\"Teriyaki Chicken Casserole\\"></div><h3>Ingredients</h3><ul class=\\"ingredientList\\"><li class=\\"listItem\\">3/4 cup&nbsp;soy sauce</li><li class=\\"listItem\\">1/2 cup&nbsp;water</li><li class=\\"listItem\\">1/4 cup&nbsp;brown sugar</li><li class=\\"listItem\\">1/2 teaspoon&nbsp;ground ginger</li><li class=\\"listItem\\">1/2 teaspoon&nbsp;minced garlic</li><li class=\\"listItem\\">4 Tablespoons&nbsp;cornstarch</li><li class=\\"listItem\\">2&nbsp;chicken breasts</li><li class=\\"listItem\\">1 (12 oz.)&nbsp;stir-fry vegetables</li><li class=\\"listItem\\">3 cups&nbsp;brown rice</li></ul><ul class=\\"ingredientImages\\"><li class=\\"listItem\\" title=\\"soy sauce\\"><img alt=\\"soy sauce\\" src=\\"https://www.themealdb.com/images/ingredients/soy sauce-Small.png\\"></li><li class=\\"listItem\\" title=\\"water\\"><img alt=\\"water\\" src=\\"https://www.themealdb.com/images/ingredients/water-Small.png\\"></li><li class=\\"listItem\\" title=\\"brown sugar\\"><img alt=\\"brown sugar\\" src=\\"https://www.themealdb.com/images/ingredients/brown sugar-Small.png\\"></li><li class=\\"listItem\\" title=\\"ground ginger\\"><img alt=\\"ground ginger\\" src=\\"https://www.themealdb.com/images/ingredients/ground ginger-Small.png\\"></li><li class=\\"listItem\\" title=\\"minced garlic\\"><img alt=\\"minced garlic\\" src=\\"https://www.themealdb.com/images/ingredients/minced garlic-Small.png\\"></li><li class=\\"listItem\\" title=\\"cornstarch\\"><img alt=\\"cornstarch\\" src=\\"https://www.themealdb.com/images/ingredients/cornstarch-Small.png\\"></li><li class=\\"listItem\\" title=\\"chicken breasts\\"><img alt=\\"chicken breasts\\" src=\\"https://www.themealdb.com/images/ingredients/chicken breasts-Small.png\\"></li><li class=\\"listItem\\" title=\\"stir-fry vegetables\\"><img alt=\\"stir-fry vegetables\\" src=\\"https://www.themealdb.com/images/ingredients/stir-fry vegetables-Small.png\\"></li><li class=\\"listItem\\" title=\\"brown rice\\"><img alt=\\"brown rice\\" src=\\"https://www.themealdb.com/images/ingredients/brown rice-Small.png\\"></li></ul><h3>Instructions</h3><ol><li class=\\"listItem\\">Preheat oven to 350° F. Spray a 9x13-inch baking pan with non-stick spray.</li><li class=\\"listItem\\">Combine soy sauce, ½ cup water, brown sugar, ginger and garlic in a small saucepan and cover. Bring to a boil over medium heat. Remove lid and cook for one minute once boiling.</li><li class=\\"listItem\\">Meanwhile, stir together the corn starch and 2 tablespoons of water in a separate dish until smooth. Once sauce is boiling, add mixture to the saucepan and stir to combine. Cook until the sauce starts to thicken then remove from heat.</li><li class=\\"listItem\\">Place the chicken breasts in the prepared pan. Pour one cup of the sauce over top of chicken. Place chicken in oven and bake 35 minutes or until cooked through. Remove from oven and shred chicken in the dish using two forks.</li><li class=\\"listItem\\">*Meanwhile, steam or cook the vegetables according to package directions.</li><li class=\\"listItem\\">Add the cooked vegetables and rice to the casserole dish with the chicken. Add most of the remaining sauce, reserving a bit to drizzle over the top when serving. Gently toss everything together in the casserole dish until combined. Return to oven and cook 15 minutes. Remove from oven and let stand 5 minutes before serving. Drizzle each serving with remaining sauce. Enjoy!</li></ol>"`
  );
});

it("renders loading message while meal is not loaded", async () => {
  mockMealId = undefined;

  await act(async () => {
    render(<Meal />, container);
  });

  expect(container!.innerHTML).toContain("Loading");
});
