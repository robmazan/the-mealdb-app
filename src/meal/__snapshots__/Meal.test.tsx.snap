// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the meal details properly 1`] = `
<Fragment>
  <Link
    className="backLink"
    to="/category/Chicken"
  >
    Back to category
  </Link>
  <h2
    className="pageTitle"
  >
    Teriyaki Chicken Casserole
  </h2>
  <div
    className="card"
  >
    <img
      alt="Teriyaki Chicken Casserole"
      className="photo"
      src="https://www.themealdb.com/images/media/meals/wvpsxx1468256321.jpg"
    />
  </div>
  <h3>
    Ingredients
  </h3>
  <ul
    className="ingredientList"
  >
    <li
      className="listItem"
      key="ing_52772_0"
    >
      3/4 cup
       
      soy sauce
    </li>
    <li
      className="listItem"
      key="ing_52772_1"
    >
      1/2 cup
       
      water
    </li>
    <li
      className="listItem"
      key="ing_52772_2"
    >
      1/4 cup
       
      brown sugar
    </li>
    <li
      className="listItem"
      key="ing_52772_3"
    >
      1/2 teaspoon
       
      ground ginger
    </li>
    <li
      className="listItem"
      key="ing_52772_4"
    >
      1/2 teaspoon
       
      minced garlic
    </li>
    <li
      className="listItem"
      key="ing_52772_5"
    >
      4 Tablespoons
       
      cornstarch
    </li>
    <li
      className="listItem"
      key="ing_52772_6"
    >
      2
       
      chicken breasts
    </li>
    <li
      className="listItem"
      key="ing_52772_7"
    >
      1 (12 oz.)
       
      stir-fry vegetables
    </li>
    <li
      className="listItem"
      key="ing_52772_8"
    >
      3 cups
       
      brown rice
    </li>
  </ul>
  <ul
    className="ingredientImages"
  >
    <li
      className="listItem"
      key="iim_52772_0"
      title="soy sauce"
    >
      <img
        alt="soy sauce"
        src="https://www.themealdb.com/images/ingredients/soy sauce-Small.png"
      />
    </li>
    <li
      className="listItem"
      key="iim_52772_1"
      title="water"
    >
      <img
        alt="water"
        src="https://www.themealdb.com/images/ingredients/water-Small.png"
      />
    </li>
    <li
      className="listItem"
      key="iim_52772_2"
      title="brown sugar"
    >
      <img
        alt="brown sugar"
        src="https://www.themealdb.com/images/ingredients/brown sugar-Small.png"
      />
    </li>
    <li
      className="listItem"
      key="iim_52772_3"
      title="ground ginger"
    >
      <img
        alt="ground ginger"
        src="https://www.themealdb.com/images/ingredients/ground ginger-Small.png"
      />
    </li>
    <li
      className="listItem"
      key="iim_52772_4"
      title="minced garlic"
    >
      <img
        alt="minced garlic"
        src="https://www.themealdb.com/images/ingredients/minced garlic-Small.png"
      />
    </li>
    <li
      className="listItem"
      key="iim_52772_5"
      title="cornstarch"
    >
      <img
        alt="cornstarch"
        src="https://www.themealdb.com/images/ingredients/cornstarch-Small.png"
      />
    </li>
    <li
      className="listItem"
      key="iim_52772_6"
      title="chicken breasts"
    >
      <img
        alt="chicken breasts"
        src="https://www.themealdb.com/images/ingredients/chicken breasts-Small.png"
      />
    </li>
    <li
      className="listItem"
      key="iim_52772_7"
      title="stir-fry vegetables"
    >
      <img
        alt="stir-fry vegetables"
        src="https://www.themealdb.com/images/ingredients/stir-fry vegetables-Small.png"
      />
    </li>
    <li
      className="listItem"
      key="iim_52772_8"
      title="brown rice"
    >
      <img
        alt="brown rice"
        src="https://www.themealdb.com/images/ingredients/brown rice-Small.png"
      />
    </li>
  </ul>
  <h3>
    Instructions
  </h3>
  <ol>
    <li
      className="listItem"
      key="ins_52772_0"
    >
      Preheat oven to 350° F. Spray a 9x13-inch baking pan with non-stick spray.
    </li>
    <li
      className="listItem"
      key="ins_52772_1"
    >
      Combine soy sauce, ½ cup water, brown sugar, ginger and garlic in a small saucepan and cover. Bring to a boil over medium heat. Remove lid and cook for one minute once boiling.
    </li>
    <li
      className="listItem"
      key="ins_52772_2"
    >
      Meanwhile, stir together the corn starch and 2 tablespoons of water in a separate dish until smooth. Once sauce is boiling, add mixture to the saucepan and stir to combine. Cook until the sauce starts to thicken then remove from heat.
    </li>
    <li
      className="listItem"
      key="ins_52772_3"
    >
      Place the chicken breasts in the prepared pan. Pour one cup of the sauce over top of chicken. Place chicken in oven and bake 35 minutes or until cooked through. Remove from oven and shred chicken in the dish using two forks.
    </li>
    <li
      className="listItem"
      key="ins_52772_4"
    >
      *Meanwhile, steam or cook the vegetables according to package directions.
    </li>
    <li
      className="listItem"
      key="ins_52772_5"
    >
      Add the cooked vegetables and rice to the casserole dish with the chicken. Add most of the remaining sauce, reserving a bit to drizzle over the top when serving. Gently toss everything together in the casserole dish until combined. Return to oven and cook 15 minutes. Remove from oven and let stand 5 minutes before serving. Drizzle each serving with remaining sauce. Enjoy!
    </li>
  </ol>
</Fragment>
`;
